### PR TITLE
feat: custom user receipt

### PR DESCRIPTION
**Pull request by hpsing**
* Custom UserOpReceiptResponse with [`reason` key ](https://github.com/eth-infinitism/bundler-spec/blob/a247b5de59a702063ea5b09d6136f119a061642b/src/schemas/receipt.yaml#L139)
* Override `GetUserOperationReceipt`
* Store `UserOpProcessingStatus` in `reason` key
* `eth_getuseroperationreceipt` rpc response with `processing status` in `reason` key

example:

1. using the solved hash

```sh
curl -X POST \
  -H "Content-Type: application/json" \
  -d '{
    "jsonrpc": "2.0",
    "method": "eth_getUserOperationReceipt",
    "params": ["0xaec7887d6c0b59c361d8dbc0a3aa6d9cd19ad26686dd26b98889a72e83e3f1d8"],
    "id": 1
  }' \
  http://localhost:4337 | jq
```

response:

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "userOpHash": "0xaec7887d6c0b59c361d8dbc0a3aa6d9cd19ad26686dd26b98889a72e83e3f1d8",
    "sender": "0xc291efdc1a6420cbb226294806604833982ed24d",
    "paymaster": "0x0000000000000000000000000000000000000000",
    "nonce": "0xaf8",
    "success": true,
    "actualGasCost": "0x2afc110400",
    "actualGasUsed": "0x963e0",
    "from": "0x43b32d92d8a6b67104d621e9a9c3831a01a8c50c",
    "receipt": {
      "blockHash": "0xca29ac53e63ac314e1b910a2559a11b5e7a96fa93e072a1299de5f95aaad9b12",
      "blockNumber": "0x1369a2f",
      "from": "0x43b32d92d8a6b67104d621e9a9c3831a01a8c50c",
      "cumulativeGasUsed": "0x536a2",
      "gasUsed": "0x536a2",
      "logs": [
        {
          "address": "0x5ff137d4b0fdcd49dca30c7cf57e578a026d2789",
          "topics": [
            "0xbb47ee3e183a558b1a2ff0874b079f3fc5478b7454eacf2bfc5af2ff5878f972"
          ],
          "data": "0x",
          "blockNumber": "0x1369a2f",
          "transactionHash": "0xa1d9c151bfdee3c8a7ebd47789661113a3653ecac02a534d1fad38fd39a2c638",
          "transactionIndex": "0x0",
          "blockHash": "0xca29ac53e63ac314e1b910a2559a11b5e7a96fa93e072a1299de5f95aaad9b12",
          "logIndex": "0x0",
          "removed": false
        },
       
      ],
      "logsBloom": "0x000000020010000000000010000010002000000000000000000000000000000000090000000000000002000100400000021000000800600000000200000000000000010000020008400000080040000000000000000000000000002080008000000800000a0000000000101000000800000000000000000000000010000800000000000000000000000000000000000000000001000000000000000000000000008010000000000000400000800001000000000000000000000002000000004001000002001000000001000200000006080000000000000000004000000020000000200000000428000000000000000400100100000000c00000000010001400",
      "transactionHash": "0xa1d9c151bfdee3c8a7ebd47789661113a3653ecac02a534d1fad38fd39a2c638",
      "transactionIndex": "0x0",
      "effectiveGasPrice": "0xc3501"
    },
    "logs": [
      {
        "address": "0x5ff137d4b0fdcd49dca30c7cf57e578a026d2789",
        "topics": [
          "0x49628fd1471006c1482da88028e9ce4dbb080b815c9b0344d39e5a8e6ec1419f",
          "0xaec7887d6c0b59c361d8dbc0a3aa6d9cd19ad26686dd26b98889a72e83e3f1d8",
          "0x000000000000000000000000c291efdc1a6420cbb226294806604833982ed24d",
          "0x0000000000000000000000000000000000000000000000000000000000000000"
        ],
        "data": "0x0000000000000000000000000000000000000000000000000000000000000af800000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000002afc11040000000000000000000000000000000000000000000000000000000000000963e0",
        "blockNumber": "0x1369a2f",
        "transactionHash": "0xa1d9c151bfdee3c8a7ebd47789661113a3653ecac02a534d1fad38fd39a2c638",
        "transactionIndex": "0x0",
        "blockHash": "0xca29ac53e63ac314e1b910a2559a11b5e7a96fa93e072a1299de5f95aaad9b12",
        "logIndex": "0xa",
        "removed": false
      }
    ],
    "reason": "PROCESSING_STATUS_ON_CHAIN"
  }
}
```

2. Using the current Hash
```sh
curl -X POST \
  -H "Content-Type: application/json" \
  -d '{
    "jsonrpc": "2.0",
    "method": "eth_getUserOperationReceipt",
    "params": ["0x2f5c4dd288455f5e0c9555c47240799dac3bbb045dc85196d93690d7210ab50b"],
    "id": 1
  }' \
  http://localhost:4337 | jq
```

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "userOpHash": "0xaec7887d6c0b59c361d8dbc0a3aa6d9cd19ad26686dd26b98889a72e83e3f1d8",
    "sender": "0xc291efdc1a6420cbb226294806604833982ed24d",
    "paymaster": "0x0000000000000000000000000000000000000000",
    "nonce": "0xaf8",
    "success": true,
    "actualGasCost": "0x2afc110400",
    "actualGasUsed": "0x963e0",
    "from": "0x43b32d92d8a6b67104d621e9a9c3831a01a8c50c",
    "receipt": {
      "blockHash": "0xca29ac53e63ac314e1b910a2559a11b5e7a96fa93e072a1299de5f95aaad9b12",
      "blockNumber": "0x1369a2f",
      "from": "0x43b32d92d8a6b67104d621e9a9c3831a01a8c50c",
      "cumulativeGasUsed": "0x536a2",
      "gasUsed": "0x536a2",
      "logs": [
        {
          "address": "0x5ff137d4b0fdcd49dca30c7cf57e578a026d2789",
          "topics": [
            "0xbb47ee3e183a558b1a2ff0874b079f3fc5478b7454eacf2bfc5af2ff5878f972"
          ],
          "data": "0x",
          "blockNumber": "0x1369a2f",
          "transactionHash": "0xa1d9c151bfdee3c8a7ebd47789661113a3653ecac02a534d1fad38fd39a2c638",
          "transactionIndex": "0x0",
          "blockHash": "0xca29ac53e63ac314e1b910a2559a11b5e7a96fa93e072a1299de5f95aaad9b12",
          "logIndex": "0x0",
          "removed": false
        },
       
      ],
      "logsBloom": "0x000000020010000000000010000010002000000000000000000000000000000000090000000000000002000100400000021000000800600000000200000000000000010000020008400000080040000000000000000000000000002080008000000800000a0000000000101000000800000000000000000000000010000800000000000000000000000000000000000000000001000000000000000000000000008010000000000000400000800001000000000000000000000002000000004001000002001000000001000200000006080000000000000000004000000020000000200000000428000000000000000400100100000000c00000000010001400",
      "transactionHash": "0xa1d9c151bfdee3c8a7ebd47789661113a3653ecac02a534d1fad38fd39a2c638",
      "transactionIndex": "0x0",
      "effectiveGasPrice": "0xc3501"
    },
    "logs": [
      {
        "address": "0x5ff137d4b0fdcd49dca30c7cf57e578a026d2789",
        "topics": [
          "0x49628fd1471006c1482da88028e9ce4dbb080b815c9b0344d39e5a8e6ec1419f",
          "0xaec7887d6c0b59c361d8dbc0a3aa6d9cd19ad26686dd26b98889a72e83e3f1d8",
          "0x000000000000000000000000c291efdc1a6420cbb226294806604833982ed24d",
          "0x0000000000000000000000000000000000000000000000000000000000000000"
        ],
        "data": "0x0000000000000000000000000000000000000000000000000000000000000af800000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000002afc11040000000000000000000000000000000000000000000000000000000000000963e0",
        "blockNumber": "0x1369a2f",
        "transactionHash": "0xa1d9c151bfdee3c8a7ebd47789661113a3653ecac02a534d1fad38fd39a2c638",
        "transactionIndex": "0x0",
        "blockHash": "0xca29ac53e63ac314e1b910a2559a11b5e7a96fa93e072a1299de5f95aaad9b12",
        "logIndex": "0xa",
        "removed": false
      }
    ],
    "reason": "PROCESSING_STATUS_ON_CHAIN"
  }
}
```

3. UserOp Failed Case (Invalid case) 
```sh
curl -X POST \
  -H "Content-Type: application/json" \
  -d '{
    "jsonrpc": "2.0",
    "method": "eth_getUserOperationReceipt",
    "params": ["0x66face11c954cc572883bd9282e505099f25ade333dbc1aa8ac29c0aea33c43b"],
    "id": 1
  }' \
  http://localhost:4337 | jq
```

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "userOpHash": "0x66face11c954cc572883bd9282e505099f25ade333dbc1aa8ac29c0aea33c43b",
    "sender": "0x0000000000000000000000000000000000000000",
    "paymaster": "0x0000000000000000000000000000000000000000",
    "nonce": "",
    "success": false,
    "actualGasCost": "",
    "actualGasUsed": "",
    "from": "0x0000000000000000000000000000000000000000",
    "receipt": null,
    "logs": null,
    "reason": "PROCESSING_STATUS_INVALID"
  }
}
```

## Summary by CodeRabbit

- **New Features**
	- Introduced functionality for managing and retrieving user operation receipts in Ethereum transactions.
	- Enhanced RPC server with custom receipt handling and improved error handling for RPC methods.
	- Added database integration for tracking user operation statuses.
	- Implemented a new storage system utilizing BadgerDB for managing user operation statuses and a whitelist of addresses.

- **Bug Fixes**
	- Improved error handling and logging throughout various components, particularly in RPC and user operation processing.

- **Documentation**
	- Minor adjustments to comments for clarity throughout the codebase.

### CODE REVIEW

File: rpc/server.go [hunk 1]
```diff
@@ -6,31 +6,134 @@ import (
...
+func (c *Client) SetGetUserOpReceiptFunc(fn receipt.GetUserOpReceiptFunc) {
+	c.getUserOpReceipt = fn
+}
+
+// GetUserOperationReceipt retrieves the receipt for a user operation.
+// It uses our custom receipt type that includes additional fields 'reason'.
+func (c *Client) GetUserOperationReceipt(hash string) (*receipt.UserOperationReceipt, error) {
+	values := conf.GetValues()
+	return c.getUserOpReceipt(hash, values.SupportedEntryPoints[0], values.OpLookupLimit)
+}
```
[the remark this answers is not in the document]
Comment (hpsing) [reply rpc/server.go:94]: not required it is always called during client initialization

File: receipt/handler.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+	var userOpStatus solution.UserOpStatusMap
+	err := db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(statusKey)
+		if err != nil {
+			if err == badger.ErrKeyNotFound {
+				return fmt.Errorf("status not found for hash: %s", userOpHash)
+			}
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (hpsing) [reply receipt/handler.go]: here, `db.View` is a closure must return a single error.

File: solution/solveintents.go [hunk 3]
```diff
@@ -51,28 +52,79 @@ type batchOpIndex int
...
 		logger:       logger,
 		txHashes:     txHashes,
 		ep:           entrypoint,
 		chainID:      chainID,
+		db:           db,
 	}
 }
 
+// UpdateProcessingStatusInDB updates the processing status in the db.
+func (ei *IntentsHandler) UpdateProcessingStatusInDB(
```
Comment (blewater) [inline solution/solveintents.go]: I suggest to add it in a new package and use it here.

File: receipt/handler.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	return ep.FilterUserOperationEvent(
+		&bind.FilterOpts{Start: startBlk.Uint64()},
+		[][32]byte{common.HexToHash(userOpHash)},
+		[]common.Address{},
+		[]common.Address{},
+	)
+}
+
+// getProcessingStatusFromDB retrieves the processing status from the database
+func getProcessingStatusFromDB(db *badger.DB, userOpHash string) (*solution.UserOpStatusMap, error) {
```
Comment (blewater) [inline receipt/handler.go]: I suggest to add it in a new package along with UpdateStatus.

**Review by blewater — CHANGES_REQUESTED** [review]
Good.
Adding new package DBStatus with 2 methods
- UpdateStatus
- GetStatus
Would allow to reuse the functionality.

File: receipt/handler.go (hunk 4 above)
Comment (hpsing) [reply receipt/handler.go]: added

**Review by blewater — DISMISSED** [review]
Great!

File: cmd/main.go [hunk 5]
```diff
@@ -81,27 +87,34 @@ func main() {
...
 	if err != nil {
 		stdLogger.Error(err, "could not open badger database")
 		os.Exit(1)
 	}
-
 	defer db.Close()
+
+	// Initialize BadgerStore with logger
+	store := store.NewBadgerStore(db, stdLogger)
+
```
[the remark this answers is not in the document]
Comment (hpsing) [reply cmd/main.go:100]: not required, as it usage pre initialized db instance

**Review by blewater — APPROVED** [review]
LGTM

**Pull request merged**